### ISSUE TITLE
Segfault after installation

### ISSUE THREAD

**Issue by Sweets** [wacossusca34/glava#19]
Freshly installed and built version of glava, running Arch linux (glfw package is `glfw-x11`).

```
Loading module: 'bars'
found GLSL stage: '1.frag'
compiling: '1.frag'
Using default PulseAudio sink: alsa_output.pci-0000_00_14.2.analog-stereo.monitor
ASAN:DEADLYSIGNAL
=================================================================
==29759==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x55a522df4844 bp 0x000000000176 sp 0x7ffcac6e1f50 T0)
==29759==The signal is caused by a READ memory access.
==29759==Hint: address points to the zero page.
    #0 0x55a522df4843 in xwin_should_render (/usr/bin/glava+0x3843)
    #1 0x55a522df40b7 in main (/usr/bin/glava+0x30b7)
    #2 0x7f2fbc46bf49 in __libc_start_main (/usr/lib/libc.so.6+0x20f49)
    #3 0x55a522df4509 in _start (/usr/bin/glava+0x3509)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/usr/bin/glava+0x3843) in xwin_should_render
==29759==ABORTING
```

**Comment by wacossusca34** [comment]
Do you always get this? Are you running on Wayland, and if not, what WM are you using?

Also, run `export ASAN_OPTIONS=symbolize=1` before running `./glava` with build instructions `make INSTALL=standalone BUILD=debug`. This gives me line numbers that I can use.

And make sure GCC is up to date.

**Comment by Sweets** [comment]
Yes, I do always get this. I don't use Wayland, and the window manager is custard ( https://github.com/Sweets/custard ). Some of my friends have tested this and successfully ran it, so I attribute it to something in my window manager (the hope is that I could get exact calls for why this happens, and then patch custard if possible).

As for running with `ASAN_OPTIONS`, and building with `INSTALL` and `BUILD`, it doesn't appear that the output is any different from before, save for memory addresses. In any case, here's the output.

```
Loading module: 'bars'
found GLSL stage: '1.frag'
compiling: '1.frag'
Using default PulseAudio sink: alsa_output.pci-0000_00_14.2.analog-stereo.monitor
ASAN:DEADLYSIGNAL
=================================================================
==32580==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x5605fd8e7844 bp 0x000000000176 sp 0x7ffce3a5bd80 T0)
==32580==The signal is caused by a READ memory access.
==32580==Hint: address points to the zero page.
    #0 0x5605fd8e7843 in xwin_should_render (/users/grace/git/glava/glava+0x3843)
    #1 0x5605fd8e70b7 in main (/users/grace/git/glava/glava+0x30b7)
    #2 0x7f1353264f49 in __libc_start_main (/usr/lib/libc.so.6+0x20f49)
    #3 0x5605fd8e7509 in _start (/users/grace/git/glava/glava+0x3509)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/users/grace/git/glava/glava+0x3843) in xwin_should_render
==32580==ABORTING
```

glibc is up to date, according to pacman. I'm not entirely positive that it is, as my system uses a mix of repositories, some of which replace the core repo itself for Arch, but the version is `glibc-2.26.11`.

**Comment by wacossusca34** [comment]
This was an uncaught bug with reading `_NET_ACTIVE_WINDOW`. Most EWMH compliant window managers have this set to _something_ (ie. `None` or `0L` if nothing is in focus), rather than leaving it empty (causing `XGetWindowProperty` to return `0` for `nitems_return`). This is technically a EWMH compliance issue on your part.

I fixed this in GLava anyway for checking that the property actually returns some items.

Make sure your WM:

* Sets the `_NET_ACTIVE_WINDOW` atom for your root X11 window
* Sets the `_NET_WM_STATE` atom to `_NET_WM_STATE_FULLSCREEN` for X11 clients if they are fullscreen.

**Comment by wacossusca34** [comment]
This is should be fixed in `unstable`, but that branch is all over the place right now. I will close this when it's merged for 1.4.

**Comment by Sweets** [comment]
Alright, I'll be updating it later today then. Thanks for the note on EWMH compliancy, by the way. :)

**Comment by Sweets** [comment]
After adding support for the `_NET_ACTIVE_WINDOW` hint, it starts successfully. It doesn't size correctly, though, ignoring the configuration altogether. custard does support _NET_WM_WINDOW_TYPE_DESKTOP, which is what I've set the xwintype to in the configuration.

**Comment by wacossusca34** [comment]
GLava _does_ set `_NET_WM_WINDOW_TYPE_DESKTOP`, and seems to work with most window managers that support it.

> After adding support for the _NET_ACTIVE_WINDOW hint, it starts successfully. It doesn't size correctly, though, ignoring the configuration altogether.

GLFW handles sizing, but it's probably done through EWMH requests as well. I haven't had anyone else report issues with GLFW resizing itself, and since most applications don't resize themselves after creation and generally leave that to the user or WM, I would put my money on your WM not handling it properly.

> but I'm not at all familiar with glfw

GLFW is just a giant layer of abstraction over various windowing systems and OpenGL context APIs. Anything I'm not explicitly doing in `xwin.c` with X11 is being handled behind the scenes with GLFW, although that will change with GLava 1.4 when I add a direct GLX backend.

I would suggest ensuring oddities you encounter with GLava are not due to the nature of your WM being under active development by testing it alongside a more stable, common WM (like openbox). You could even run multiple WMs with [Xephyr](https://en.wikipedia.org/wiki/Xephyr).

**Comment by Sweets** [comment]
I previously did development exclusively in Xephyr, actually. I don't have it installed _now_ because I just did a fresh install, that is supposed to be minimal. Anyways,

`_NET_WM_WINDOW_TYPE_DESKTOP` is properly supported. Also, perhaps I worded that badly. It's not that it resizes, it's that it doesn't use the size specified in configuration, and I do agree that it's likely the window manager.

But here's what stumps me. Whenever the window manager checks the EWMH hints of a window, if it sees that a window is of `_NET_WM_WINDOW_TYPE_DESKTOP`, it will simply not manage it. Relevant line is here: https://github.com/Sweets/custard/blob/master/src/window.cpp#L23

To support this, it's even shown in windows as well. The way the WM works, is that windows are sized and moved across coordinates on a virtual grid. Whenever a window is opened, if the window manager deems that it should be managed, then it will resize it to be in the center of the virtual grid. Relevant here: https://github.com/Sweets/custard/blob/master/src/window.cpp#L82

That being said, here's an example of what exactly is happening when starting GLava. 
![23 10 24_full](https://user-images.githubusercontent.com/13605846/36084215-cb385834-0f80-11e8-9d16-b684c9d7fa83.png)

The window in the center is a freshly opened terminal that has not been moved or resized. Top left is GLava, which isn't managed by the WM. This screenshot was taken directly after GLava was opened.
